packages/liveblocks-core/src/lib/fsm.ts

```diff
@@ -128,6 +125,58 @@ function patterns<TState extends string>(
   return result;
 }
 
+class SafeContext<TContext extends object> {
+  private curr: Readonly<TContext>;
+
+  constructor(initialContext: TContext) {
+    this.curr = initialContext;
+  }
+
+  get current(): Readonly<TContext> {
+    return this.curr;
+  }
+
+  /**
+   * Call a callback function that allows patching of the context, by
+   * calling `context.patch()`. Patching is only allowed for the duration
+   * of this window.
+   */
+  allowPatching(callback: (context: Patchable<TContext>) => void): void {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+    let allowed = true;
+
+    const patchableContext = {
+      ...this.curr,
+      patch(patch: Partial<TContext>): void {
+        if (allowed) {
+          self.curr = Object.assign({}, self.curr, patch);
+
+          // Also patch the temporary mutable context helper itself, in case
+          // there are multiple calls in a succession that need
+          for (const pair of Object.entries(patch)) {
+            const [key, value] = pair as [
+              keyof TContext,
+              TContext[keyof TContext]
+            ];
+            if (key !== "patch") {
```

Comment:
I'm not sure about this limitation, what if someone actually does want to patch a prop named patch?

Reply:
It would not be possible. I thought about it, but chose pragmatism over purity, because otherwise we'd have to send it as two values into each callback, requiring unpacking, i.e. `(ctx, event, patch) => patch({ … })` or `({ ctx, patch }, event) => patch({ … })`, both of which feels overly complicated.

If ever the need for a `context.patch` property arises, I'd be happy to reconsider this API.